File: LeetcodeSwift/LeetcodeSwift/algorithms/q0146.swift
//
//  q0146.swift
//  LeetcodeSwift
//
//  Created by NowOrNever on 24/04/2020.
//  Copyright © 2020 DL. All rights reserved.
//

import Foundation
//146. LRU Cache
//Medium
//
//5052
//
//227
//
//Add to List
//
//Share
//Design and implement a data structure for Least Recently Used (LRU) cache. It should support the following operations: get and put.
//
//get(key) - Get the value (will always be positive) of the key if the key exists in the cache, otherwise return -1.
//put(key, value) - Set or insert the value if the key is not already present. When the cache reached its capacity, it should invalidate the least recently used item before inserting a new item.
//
//The cache is initialized with a positive capacity.
//
//Follow up:
//Could you do both operations in O(1) time complexity?
//
//Example:
//
//LRUCache cache = new LRUCache( 2 /* capacity */ );
//
//cache.put(1, 1);
//cache.put(2, 2);
//cache.get(1);       // returns 1
//cache.put(3, 3);    // evicts key 2
//cache.get(2);       // returns -1 (not found)
//cache.put(4, 4);    // evicts key 1
//cache.get(1);       // returns -1 (not found)
//cache.get(3);       // returns 3
//cache.get(4);       // returns 4
//
//
//Accepted
//476,364
//Submissions
//1,539,139

class LRUCache {
    var dic = [Int:Int]()
    var dicCount = [Int:Int]()
    var stack1 = [Int]()
    var capacity = 0

    init(_ capacity: Int) {
        self.capacity = capacity
    }
    
    func get(_ key: Int) -> Int {
        if dic[key] != nil {
            stack1.append(key)
            dicCount[key]! += 1
            return dic[key]!
        }
        return -1
    }
    
    func put(_ key: Int, _ value: Int) {
        if dic[key] == nil && dic.count >= capacity{
            while dicCount[stack1.first!]! > 1 {
                dicCount[stack1.first!]! -= 1
                stack1.removeFirst()
            }
            dicCount.removeValue(forKey: stack1.first!)
            dic.removeValue(forKey: stack1.first!)
            stack1.removeFirst()
        }
        stack1.append(key)
        dic[key] = value
        dicCount[key] = (dicCount[key] ?? 0) + 1
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * let obj = LRUCache(capacity)
 * let ret_1: Int = obj.get(key)
 * obj.put(key, value)
 */
func q0146() {
    let cache = LRUCache.init(2)
    cache.put(2, 1)
    cache.put(3, 2)
    print(cache.get(3))
    print(cache.get(2))
    cache.put(4, 3)
    print(cache.get(2))
    print(cache.get(3))
    print(cache.get(4))
}

File: LeetcodePython3/q0146.py
class DoubleListNode:
    def __init__(self, key=0, val=0, next=None, pre=None):
        self.key: int = key
        self.val: int = val
        self.next = next
        self.pre = pre


class LRUCache:
    capacity = 0
    cache = {}
    head = None
    tail = None
    size = 0

    def __init__(self, capacity: int):
        self.capacity = capacity
        self.cache = {}
        self.head = DoubleListNode(0)
        self.tail = DoubleListNode(0)
        self.head.next = self.tail
        self.tail.pre = self.head
        self.size = 0

    def get(self, key: int) -> int:
        if key not in self.cache:
            return -1
        else:
            node = self.cache[key]
            self.moveTohead(node)
            return node.val

    def put(self, key: int, value: int) -> None:
        if key in self.cache:
            node = self.cache[key]
            node.val = value
            self.moveTohead(node)
        else:
            node = DoubleListNode(key, value)
            self.cache[key] = node
            self.addToHead(node)
            self.size += 1
            if self.size > self.capacity:
                remove_key = self.removeTail().key
                self.cache.pop(remove_key)
                self.size -= 1

    def deleteNode(self, node: DoubleListNode):
        node.pre.next = node.next
        node.next.pre = node.pre

    def addToHead(self, node: DoubleListNode):
        node.pre = self.head
        node.next = self.head.next
        self.head.next.pre = node
        self.head.next = node

    def removeTail(self) -> DoubleListNode:
        node = self.tail.pre
        self.deleteNode(node)
        return node

    def moveTohead(self, node: DoubleListNode):
        self.deleteNode(node)
        self.addToHead(node)

    # Your LRUCache object will be instantiated and called as such:
capacity = 2
obj = LRUCache(capacity)
obj.put(1,1)
obj.put(2,2)
result = obj.get(1)
print(result)

# param_1 = obj.get(key)
# obj.put(key,value)
